gson/src/main/java/com/google/gson/Gson.java

```diff
@@ -523,47 +532,55 @@ public <T> TypeAdapter<T> getAdapter(TypeToken<T> type) {
       return adapter;
     }
 
-    Map<TypeToken<?>, FutureTypeAdapter<?>> threadCalls = calls.get();
-    boolean requiresThreadLocalCleanup = false;
+    Map<TypeToken<?>, TypeAdapter<?>> threadCalls = threadLocalAdapterResults.get();
+    boolean isInitialAdapterRequest = false;
     if (threadCalls == null) {
       threadCalls = new HashMap<>();
-      calls.set(threadCalls);
-      requiresThreadLocalCleanup = true;
-    }
-
-    // the key and value type parameters always agree
-    @SuppressWarnings("unchecked")
-    FutureTypeAdapter<T> ongoingCall = (FutureTypeAdapter<T>) threadCalls.get(type);
-    if (ongoingCall != null) {
-      return ongoingCall;
+      threadLocalAdapterResults.set(threadCalls);
+      isInitialAdapterRequest = true;
+    } else {
```

Comment:
Have moved the code here in the `else` block; otherwise the `ongoingCall` check will be unsuccessful since the `if` statement just created an empty map as `threadCalls` value.